ui/vty/src/Ariadne/UI/Vty/Widget.hs

```diff
@@ -0,0 +1,394 @@
+  Widget widget@WidgetInfo{..} <- get
+  handleByName widget name
+    (widgetHandlePaste pasted)
+    (handleWidgetPaste pasted)
+
+handleWidgetMouseDown :: B.Location -> WidgetName -> StateT (Widget p) (StateT p (B.EventM WidgetName)) WidgetEventResult
+handleWidgetMouseDown coords name = do
+  Widget widget@WidgetInfo{..} <- get
+  handleByName widget name
+    (widgetHandleMouseDown coords)
+    (handleWidgetMouseDown coords)
+
+handleWidgetScroll :: ScrollingAction -> WidgetName -> StateT (Widget p) (StateT p (B.EventM WidgetName)) WidgetEventResult
+handleWidgetScroll action name = do
+  Widget widget@WidgetInfo{..} <- get
+  handleByName widget name
+    (widgetHandleScroll action)
+    (handleWidgetScroll action)
+
+handleWidgetEvent :: UiEvent -> StateT (Widget p) (StateT p (B.EventM WidgetName)) ()
+handleWidgetEvent event = do
+  Widget widget@WidgetInfo{..} <- get
+  handleAll widget
+    (widgetHandleEvent event)
+    (handleWidgetEvent event)
+
+----------------------------------------------------------------------------
+-- Event handling helpers
+----------------------------------------------------------------------------
+
+handleAll
+  :: WidgetInfo s p
+  -> WidgetEventM s p ()
+  -> StateT (Widget (WidgetInfo s p)) (StateT (WidgetInfo s p) (B.EventM WidgetName)) ()
+  -> StateT (Widget p) (StateT p (B.EventM WidgetName)) ()
+handleAll widget@WidgetInfo{..} widgetHandler childHandler = do
+  (widgetChildren', widget') <- lift . lift $ runStateT (mapM (execStateT childHandler) widgetChildren) widget
+  withWidget widget'{ widgetChildren = widgetChildren' } widgetHandler
+
+handleByName
+  :: WidgetInfo s p
+  -> WidgetName
+  -> WidgetEventM s p WidgetEventResult
+  -> (WidgetName -> StateT (Widget (WidgetInfo s p)) (StateT (WidgetInfo s p) (B.EventM WidgetName)) WidgetEventResult)
+  -> StateT (Widget p) (StateT p (B.EventM WidgetName)) WidgetEventResult
+handleByName widget@WidgetInfo{..} name widgetHandler childHandler =
+  case name of
+    [] -> withWidget widget widgetHandler
+    x:xs -> case Map.lookup x widgetChildren of
+      Nothing -> withWidget widget widgetHandler
+      Just child -> do
+        ((res, child'), widget') <- lift . lift $ runStateT (runStateT (childHandler xs) child) widget
+        withWidget widget'{ widgetChildren = Map.insert x child' widgetChildren } $
+          case res of
+            WidgetEventNotHandled -> widgetHandler
+            WidgetEventHandled -> return res
+
+withWidget
+  :: WidgetInfo s p
+  -> WidgetEventM s p a
+  -> StateT (Widget p) (StateT p (B.EventM WidgetName)) a
+withWidget widget@WidgetInfo{..} action = do
+  let
+    subaction = do
+      (res, widgetState') <- runStateT action widgetState
+      widgetStateL .= widgetState'
+      forM_ widgetEventQueue $ \(namePart, event) -> do
+        whenJust (Map.lookup namePart widgetEventHandlers) $ \handler -> do
+          use widgetStateL >>= execStateT (handler event) >>= assign widgetStateL
```

Comment:
`widgetStateL %= execStateT (handler event)` ?

Reply:
Yeah, probably

Reply:
Nope, second argument of `%=` must be `s -> s`, but `execStateT (handler event)` is `s -> m s`

Reply:
Okay, fair.